copy.go

```diff
@@ -54,6 +57,83 @@ type CopyOptions struct {
 	MapRoot func(ctx context.Context, src content.Storage, root ocispec.Descriptor) (ocispec.Descriptor, error)
 }
 
+// getPlatformFromConfig returns a platform object which is made up from the
+// fields in config blob.
+func getPlatformFromConfig(ctx context.Context, src content.Storage, desc ocispec.Descriptor, targetConfigMediaType string) (*ocispec.Platform, error) {
+	if desc.MediaType != targetConfigMediaType {
+		return nil, fmt.Errorf("mismatch MediaType %s: expect %s", desc.MediaType, targetConfigMediaType)
+	}
+
+	rc, err := src.Fetch(ctx, desc)
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	var platform ocispec.Platform
+	if err = json.NewDecoder(rc).Decode(&platform); err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	return &platform, nil
+}
+
+// selectPlatform implements platform filter and returns the descriptor of the
+// first matched manifest if the root is a manifest list. If the root is a
+// manifest, then return the root descriptor if platform matches.
+func selectPlatform(ctx context.Context, src content.Storage, root ocispec.Descriptor, p *ocispec.Platform) (ocispec.Descriptor, error) {
+	switch root.MediaType {
+	case docker.MediaTypeManifestList, ocispec.MediaTypeImageIndex:
+		manifests, err := content.Successors(ctx, src, root)
+		if err != nil {
+			return ocispec.Descriptor{}, err
+		}
+
+		// platform filter
+		for _, m := range manifests {
+			if platform.Match(m.Platform, p) {
+				return m, nil
+			}
+		}
+		return ocispec.Descriptor{}, errdef.ErrNotFound
+	case docker.MediaTypeManifest, ocispec.MediaTypeImageManifest:
+		descs, err := content.Successors(ctx, src, root)
+		if err != nil {
+			return ocispec.Descriptor{}, err
+		}
+
+		configMediaType := docker.MediaTypeConfig
+		if root.MediaType == ocispec.MediaTypeImageManifest {
+			configMediaType = ocispec.MediaTypeImageConfig
+		}
+
+		cfgPlatform, err := getPlatformFromConfig(ctx, src, descs[0], configMediaType)
+		if err != nil {
+			return ocispec.Descriptor{}, err
+		}
+
+		if platform.Match(cfgPlatform, p) {
+			return root, nil
+		}
+		return ocispec.Descriptor{}, errdef.ErrNotFound
+	default:
+		return ocispec.Descriptor{}, fmt.Errorf("%s: %s: %w", root.Digest, root.MediaType, errdef.ErrUnsupported)
+	}
+}
+
+// WithTargetPlatform adds the check on the platform attributes.
```

Comment:
Should we elaborate the doc a bit more? I think the below information can be helpful:
1. What does this function do
2. If `MapRoot` is provided, the platform selection will be applied on the mapped root

Reply:
How about this
```
// WithTargetPlatform adds the check on the platform attributes. If MapRoot is
// provided, the platform selection will be applied on the mapped root.
// - If the root is a manifest, it will remain the same if platform matches,
// otherwise the root will be nil.
// - If the root is a manifest list, it will be reassigned to the first manifest
// with matching platform attributes.
func (o *CopyOptions) WithTargetPlatform(p *ocispec.Platform) {
```

Reply:
I feel that documenting is hard... 
Some thoughts:
> ... otherwise the root will be nil.

How about:
> ... otherwise ErrNotFound will be returned.

?

> WithTargetPlatform adds the check on the platform attributes.

This looks not very matching the function name😂

> If the root is a manifest list, it will be reassigned to the first manifest

root will be reassigned to ...? It might be a little bit strange, but I don't have a good idea at this moment😂

Reply:
yes it is hard😂
How about:
```
// WithTargetPlatform configures opts.MapRoot to select the manifest whose 
// platform matches the given platform. If MapRoot is provided, the platform 
// selection will be applied on the mapped root.
// - If the root is a manifest, it will remain the same if platform matches,
// otherwise ErrNotFound will be returned.
// - If the root is a manifest list, it will be changed to the first matching 
// manifest. If there is no matching manifest, will return ErrNotFound.
func (opts *CopyOptions) WithTargetPlatform(p *ocispec.Platform) {
```